source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -368,6 +368,70 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
   }
 }
 
+void ContextImpl::populateServerNamesMap(TlsContext& ctx, const int pkey_id) {
+  std::vector<std::string> server_name_patterns;
+  if (ctx.cert_chain_ == nullptr) {
+    return;
+  }
+  bssl::UniquePtr<GENERAL_NAMES> san_names(static_cast<GENERAL_NAMES*>(
+      X509_get_ext_d2i(ctx.cert_chain_.get(), NID_subject_alt_name, nullptr, nullptr)));
+  auto dns_sans = Utility::getSubjectAltNames(*ctx.cert_chain_, GEN_DNS);
+  // https://www.rfc-editor.org/rfc/rfc6066#section-3
+  // Currently, the only server names supported are DNS hostnames, so we
+  // only save dns san entries to match SNI.
+  for (const auto& san : dns_sans) {
+    if (absl::StartsWith(san, "*.")) {
+      server_name_patterns.emplace_back(san.substr(1));
+    } else {
+      server_name_patterns.emplace_back(san);
+    }
+  }
+
+  // https://www.rfc-editor.org/rfc/rfc6125#section-6.4.4
+  // As noted, a client MUST NOT seek a match for a reference identifier
+  // of CN-ID if the presented identifiers include a DNS-ID, SRV-ID,
+  // URI-ID, or any application-specific identifier types supported by the
+  // client.
+  if (san_names.get() == nullptr) {
+    X509_NAME* cert_subject = X509_get_subject_name(ctx.cert_chain_.get());
+    const int cn_index = X509_NAME_get_index_by_NID(cert_subject, NID_commonName, -1);
+    if (cn_index >= 0) {
+      X509_NAME_ENTRY* cn_entry = X509_NAME_get_entry(cert_subject, cn_index);
+      if (cn_entry) {
+        ASN1_STRING* cn_asn1 = X509_NAME_ENTRY_get_data(cn_entry);
+        if (cn_asn1) {
+          const auto& subject_cn =
+              std::string(reinterpret_cast<char const*>(ASN1_STRING_data(cn_asn1)));
+          if (absl::StartsWith(subject_cn, "*.")) {
+            server_name_patterns.emplace_back(subject_cn.substr(1));
+          } else {
+            server_name_patterns.emplace_back(subject_cn);
+          }
+        }
+      }
+    }
+  }
+
+  for (auto& server_name_pattern : server_name_patterns) {
+    auto sn_match = server_names_map_.find(server_name_pattern);
+    if (sn_match != server_names_map_.end()) {
+      auto pt_match = sn_match->second.find(pkey_id);
+      if (pt_match != sn_match->second.end()) {
+        throw EnvoyException(fmt::format(
+            "Failed to load certificate chain from {}, at most one "
```

Comment:
Please include the name that is conflicting in the error message. If many certs are present, it could be very difficult to figure out which name is conflicting.

Reply:
done